subsys/net/lib/http/http_app_client.c

```diff
@@ -0,0 +1,720 @@
+/*
+ * Copyright (c) 2017 Intel Corporation.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#if defined(CONFIG_NET_DEBUG_HTTP)
+#if defined(CONFIG_HTTPS)
+#define SYS_LOG_DOMAIN "https/client"
+#else
+#define SYS_LOG_DOMAIN "http/client"
+#endif
+#define NET_SYS_LOG_LEVEL SYS_LOG_LEVEL_DEBUG
+#define NET_LOG_ENABLED 1
+#endif
+
+#include <zephyr.h>
+#include <string.h>
+#include <strings.h>
+#include <errno.h>
+#include <stdlib.h>
+#include <version.h>
+
+#include <net/net_core.h>
+#include <net/net_ip.h>
+#include <net/http.h>
+
+#include "../../ip/net_private.h"
+
+#define BUF_ALLOC_TIMEOUT 100
+
+#define RC_STR(rc)	(rc == 0 ? "OK" : "ERROR")
+
+#define HTTP_EOF           "\r\n\r\n"
+
+#define HTTP_HOST          "Host"
+#define HTTP_CONTENT_TYPE  "Content-Type"
+#define HTTP_CONTENT_LEN   "Content-Length"
+#define HTTP_CONT_LEN_SIZE 6
+
+int client_reset(struct http_ctx *ctx)
+{
+	http_parser_init(&ctx->http.parser, HTTP_RESPONSE);
+
+	memset(ctx->http.rsp.http_status, 0,
+	       sizeof(ctx->http.rsp.http_status));
+
+	ctx->http.rsp.cl_present = 0;
+	ctx->http.rsp.content_length = 0;
+	ctx->http.rsp.processed = 0;
+	ctx->http.rsp.body_found = 0;
+	ctx->http.rsp.message_complete = 0;
+	ctx->http.rsp.body_start = NULL;
+
+	memset(ctx->http.rsp.response_buf, 0, ctx->http.rsp.response_buf_len);
+	ctx->http.rsp.data_len = 0;
+
+	return 0;
+}
+
+int http_request(struct http_ctx *ctx, struct http_request *req, s32_t timeout)
+{
+	const char *method = http_method_str(req->method);
+	int ret;
+
+	if (ctx->pending) {
+		net_pkt_unref(ctx->pending);
+		ctx->pending = NULL;
+	}
+
+	http_add_header(ctx, method, ctx->http.req.user_data);
```

Comment:
Either error handling, or comment like "Here we expect that we can put a bunch of stuff into freshly allocated pkt (without it being sent, which can lead to I/O error), even though nothing in TCP spec says we could."

Yeah, that's why they invented exception handling... I personally try to wrap that into macros whenever possible, IMHO it's very cumbersome to do it otherwise.

Reply:
Note that (besides exception handling) another way to deal with it is to have an explicit buffer to construct request (not relying on the fact that we can allocate given amount of data in a pkt, because we can't - MSS can be negotiated to be 1).